File: insightpdf/ai.py
from __future__ import annotations

import json
import os

from openai import OpenAI

from .models import ExtractedPage, ReportContent, Section


def create_report_content(page: ExtractedPage) -> ReportContent:
    if not os.getenv("OPENAI_API_KEY"):
        return _fallback_report(page)

    client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
    fallback = _fallback_report(page)

    try:
        response = client.chat.completions.create(
            model="gpt-4o-mini",
            temperature=0.2,
            response_format={"type": "json_object"},
            messages=[
                {
                    "role": "system",
                    "content": "Create concise business-friendly webpage report content. Return valid JSON only.",
                },
                {
                    "role": "user",
                    "content": json.dumps(
                        {
                            "instructions": "Return summary, keyInsights array, and sections array with label/body. Do not invent facts.",
                            "url": page.url,
                            "title": page.title,
                            "description": page.description,
                            "headings": page.headings,
                            "searchResults": [link.to_dict() for link in page.search_results],
                            "moreInfoLinks": [link.to_dict() for link in page.more_info_links()],
                            "text": page.text,
                        }
                    ),
                },
            ],
        )
        parsed = json.loads(response.choices[0].message.content or "{}")
    except Exception:
        return fallback

    summary = _as_text(parsed.get("summary")) or fallback.summary
    key_insights = [_as_text(item) for item in parsed.get("keyInsights", []) if _as_text(item)][:6]
    sections = [
        Section(label=_as_text(item.get("label")), body=_as_text(item.get("body")))
        for item in parsed.get("sections", [])
        if isinstance(item, dict) and _as_text(item.get("label")) and _as_text(item.get("body"))
    ][:4]

    return ReportContent(
        summary=summary,
        key_insights=key_insights or fallback.key_insights,
        sections=sections or fallback.sections,
        ai_enabled=True,
    )


def _fallback_report(page: ExtractedPage) -> ReportContent:
    sentences = [sentence.strip() for sentence in page.text.replace("?", ".").replace("!", ".").split(".")]
    key_insights = [sentence for sentence in sentences if len(sentence) > 45][:5]
    if not key_insights:
        key_insights = ["InsightPDF found public information that can be used for a compact report."]

    more_links = page.more_info_links()
    sections = [
        Section("Source Overview", page.description or "The source page was analyzed for public visible information."),
        Section(
            "Main Topics",
            (
                f"Top public search results include {', '.join(link.label for link in page.search_results[:5])}."
                if page.search_results
                else ", ".join(page.headings[:6]) or "The page has limited headings, so the report uses available text and links."
            ),
        ),
        Section(
            "Find More Information",
            "; ".join(f"{index + 1}. {link.label} ({link.href})" for index, link in enumerate(more_links))
            or "No additional public links were available.",
        ),
        Section(
            "Content Notes",
            (
                "This page exposed limited readable body text, so InsightPDF used metadata, headings, and links."
                if page.low_text
                else f"InsightPDF extracted roughly {page.word_count:,} words from the public webpage."
            ),
        ),
    ]

    return ReportContent(
        summary=_fallback_summary(page.text),
        key_insights=key_insights,
        sections=sections,
        ai_enabled=False,
    )


def _fallback_summary(text: str) -> str:
    words = text.split()
    if not words:
        return "InsightPDF generated a compact report from the available public page signals."
    return " ".join(words[:95]) + ("..." if len(words) > 95 else "")


def _as_text(value) -> str:
    return value.strip() if isinstance(value, str) else ""

File: lib/ai.ts
import OpenAI from "openai";
import type { ExtractedPage } from "@/lib/extract";

export type AiReport = {
  summary: string;
  keyInsights: string[];
  sections: {
    label: string;
    body: string;
  }[];
  aiEnabled: boolean;
};

export async function createAiReport(page: ExtractedPage, url: string): Promise<AiReport> {
  if (!process.env.OPENAI_API_KEY) {
    return createFallbackReport(page);
  }

  const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY
  });

  const completion = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    temperature: 0.2,
    response_format: { type: "json_object" },
    messages: [
      {
        role: "system",
        content:
          "You turn webpage text into concise business-friendly PDF report content. Return valid JSON only."
      },
      {
        role: "user",
        content: JSON.stringify({
          instructions:
            "Create a concise report with summary, 5 key insights, and 3 organized sections. Avoid invented facts.",
          url,
          title: page.title,
          description: page.description,
          headings: page.headings,
          text: page.text
        })
      }
    ]
  });

  const raw = completion.choices[0]?.message.content;
  if (!raw) {
    return createFallbackReport(page);
  }

  try {
    const parsed = JSON.parse(raw) as Partial<AiReport>;
    return {
      summary: ensureString(parsed.summary) || createFallbackSummary(page.text),
      keyInsights: ensureStringArray(parsed.keyInsights).slice(0, 6),
      sections: ensureSections(parsed.sections).slice(0, 4),
      aiEnabled: true
    };
  } catch {
    return createFallbackReport(page);
  }
}

function createFallbackReport(page: ExtractedPage): AiReport {
  const sentences = page.text.match(/[^.!?]+[.!?]+/g) ?? [page.text];
  const keyInsights = sentences
    .map((sentence) => sentence.trim())
    .filter((sentence) => sentence.length > 45)
    .slice(0, 5);

  return {
    summary: createFallbackSummary(page.text),
    keyInsights:
      keyInsights.length > 0
        ? keyInsights
        : ["The webpage contains extractable public content ready for report generation."],
    sections: [
      {
        label: "Source Overview",
        body: page.description || "The source page was analyzed for its visible public content."
      },
      {
        label: "Main Topics",
        body:
          page.headings.slice(0, 6).join(", ") ||
          "The page did not expose many clear headings, so the report focuses on body text."
      },
      {
        label: "Content Notes",
        body: `InsightPDF extracted roughly ${page.wordCount.toLocaleString()} words from the public webpage.`
      }
    ],
    aiEnabled: false
  };
}

function createFallbackSummary(text: string) {
  const words = text.split(/\s+/).slice(0, 95).join(" ");
  return `${words}${text.split(/\s+/).length > 95 ? "..." : ""}`;
}

function ensureString(value: unknown) {
  return typeof value === "string" ? value.trim() : "";
}

function ensureStringArray(value: unknown) {
  return Array.isArray(value)
    ? value.map((item) => ensureString(item)).filter(Boolean)
    : [];
}

function ensureSections(value: unknown) {
  if (!Array.isArray(value)) {
    return [];
  }

  return value
    .map((item) => ({
      label: ensureString((item as { label?: unknown }).label),
      body: ensureString((item as { body?: unknown }).body)
    }))
    .filter((item) => item.label && item.body);
}
